File: 159-longest-substring-with-at-most-two-distinct-characters/159.py
class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s: str) -> int:
        if len(s) <= 2:
            return len(s)

        ans = 0
        head = 0
        tail = 1
        length = 2
        table = {s[0]: 1}

        while tail < len(s):
            if s[tail] in table:
                table[s[tail]] += 1
                tail += 1
                continue

            elif len(table) == 1:
                table[s[tail]] = 1
                tail += 1
                continue

            else:
                ans = max(ans, tail-head)
                while head < tail:
                    table[s[head]] -= 1
                    if table[s[head]] == 0:
                        table.pop(s[head], None)
                    head += 1
                    if len(table) == 1:
                        break

        ans = max(ans, tail-head)
        return ans

File: 159-longest-substring-with-at-most-two-distinct-characters/159.cpp
class Solution {
public:
    int lengthOfLongestSubstringTwoDistinct(string s) {
        unordered_map<char, int> m;
        int i = 0;
        int j = 0;
        int maximum = 0;

        while (j < s.size()) {
            ++m[s[j]];
            if (m.size() > 2) {
                maximum = max(maximum, j - i);
                while (i < j && m.size() > 2) {
                    --m[s[i]];
                    if (m[s[i]] == 0)
                        m.erase(s[i]);
                    ++i;
                }
            }
            ++j;
        }

        return max(maximum, j - i);
    }
};
